### PR TITLE
[ET-1145] Feature/et 1145 add export button attendees title

### PR DESCRIPTION
**Pull request by redscar**
### 🎫 Ticket

[ET-1145] 

### 🗒️ Description

- Added the export button next to the page title on the Attendees page
- Updated the action `tribe_report_page_after_text_label` to include `$attendees` variable
- Moved the export button generating script to it's own function so it can be reused

### 🎥 Artifacts 

Showing export button next to page title
image
Showing export button on table
image
Showing no export button with no attendees
image

### ✔️ Checklist
- [x] I've included a changelog entry both in readme.txt and changelog.txt files. 
- [x] My code is tested. 
- [x] My code has [proper inline documentation](https://the-events-calendar.github.io/products-engineering/docs/code-standards/).

[ET-1145]: https://theeventscalendar.atlassian.net/browse/ET-1145

### CODE REVIEW

File: readme.txt [hunk 1]
```diff
@@ -179,8 +179,9 @@ Check out our extensive [knowledgebase](https://evnt.is/18wm) for articles on us
 == Changelog ==
 
 = [5.2.1] TBD =
-
-* Language -
+6
```
Comment (juanfra) [inline readme.txt:182]: This `6` should be an empty line.

File: src/Tribe/Attendees.php [hunk 2]
```diff
@@ -43,6 +44,8 @@ public function hook() {
 
 		add_filter( 'post_row_actions', array( $this, 'filter_admin_row_actions' ) );
 		add_filter( 'page_row_actions', array( $this, 'filter_admin_row_actions' ) );
+
+
```
Comment (juanfra) [inline src/Tribe/Attendees.php:48]: We shouldn't be introducing blank lines here.

File: src/Tribe/Attendees.php [hunk 3]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
 		return $provider->update_attendee( $attendee, $attendee_data );
 	}
 
+	/**
+	 * Generate the export URL for exporting attendees
+	 *
+	 * @since TBD
+	 *
```
Comment (juanfra) [inline src/Tribe/Attendees.php:945]: Let's remove one empty line here.

File: src/Tribe/Attendees.php [hunk 4]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
 		return $provider->update_attendee( $attendee, $attendee_data );
 	}
 
+	/**
+	 * Generate the export URL for exporting attendees
```
Comment (juanfra) [inline src/Tribe/Attendees.php:942]: The docblock must have a full stop
```suggestion
	 * Generate the export URL for exporting attendees.
```

File: src/Tribe/Attendees.php [hunk 5]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
...
+	 *
+	 * @since TBD
+	 *
+	 *
+	 * @return string Relative URL for the export.
+	 */
+	public function get_export_url (){
```
Comment (juanfra) [inline src/Tribe/Attendees.php:949]:
```suggestion
	public function get_export_url() {
```
We need to fix the spacing here.

File: src/Tribe/Attendees.php [hunk 6]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
...
+			]
+		);
+	}
+	/**
+	 * Echo the button for the export that appears next to the attendees page title.
+	 *
+	 * @since TBD
+	 *
+	 * @param int $event_id
+	 * @param Tribe__Tickets__Attendees $attendees
```
Comment (juanfra) [inline src/Tribe/Attendees.php:963]: Let's add a description for these two parameters.

File: src/Tribe/Attendees.php [hunk 7]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
...
+	 * @return string Relative URL for the export.
+	 */
+	public function get_export_url (){
+		return  add_query_arg(
+			[
+				'attendees_csv'       => true,
+				'attendees_csv_nonce' => wp_create_nonce( 'attendees_csv_nonce' ),
+			]
+		);
+	}
```
Comment (juanfra) [inline src/Tribe/Attendees.php:956]: Let's have an empty line between the closure and the docblock.

File: src/Tribe/Attendees.php [hunk 8]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
...
+	 * @since TBD
+	 *
+	 * @param int $event_id
+	 * @param Tribe__Tickets__Attendees $attendees
+	 *
+	 * @return string Relative URL for the export.
+	 */
+	public function include_export_button_title( $event_id, Tribe__Tickets__Attendees $attendees ){
+
+		// Bail early if there are no attendees
```
Comment (juanfra) [inline src/Tribe/Attendees.php:969]: Comments must end with full stop.

File: src/Tribe/Attendees.php [hunk 9]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
...
+		if ( ! $attendees->attendees_table->has_items() ) {
+			return;
+		}
+
+		// Bail early if user is not owner/have permissions
```
Comment (juanfra) [inline src/Tribe/Attendees.php:974]: Comments must end with full stop.

File: src/Tribe/Attendees.php [hunk 10]
```diff
@@ -935,4 +938,45 @@ public function update_attendee( $attendee, $attendee_data ) {
...
+		if ( ! $this->user_can_manage_attendees( 0, $event_id ) ) {
+			return;
+		}
+
+		echo sprintf( '<a target="_blank" href="%s" class="export action page-title-action">%s</a>', esc_url( $export_url = $this->get_export_url() ), esc_html__( 'Export', 'event-tickets' ) ) ;
```
Comment (juanfra) [inline src/Tribe/Attendees.php:979]: For security issues we must have `rel="noopener noreferrer"` when the `target` is `_blank`

Comment (juanfra) [reply src/Tribe/Attendees.php:979]: We could possibly split this into multiple lines for better readability.

File: src/Tribe/Attendees_Table.php [hunk 11]
```diff
@@ -602,7 +595,7 @@ public function extra_tablenav( $which ) {
 		$nav = [
 			'left' => [
 				'print'  => sprintf( '<input type="button" name="print" class="print button action" value="%s">', esc_attr__( 'Print', 'event-tickets' ) ),
-				'export' => sprintf( '<a target="_blank" href="%s" class="export button action">%s</a>', esc_url( $export_url ), esc_html__( 'Export', 'event-tickets' ) ),
+				'export' => sprintf( '<a target="_blank" href="%s" class="export button action">%s</a>', esc_url( tribe( 'tickets.attendees' )->get_export_url() ), esc_html__( 'Export', 'event-tickets' ) ),
```
Comment (juanfra) [inline src/Tribe/Attendees_Table.php:598]: For security issues we must have `rel="noopener noreferrer"` when the `target` is `_blank`

**Review by juanfra — CHANGES_REQUESTED** [review]

File: src/Tribe/Attendees.php [hunk 12]
```diff
@@ -935,4 +936,48 @@ public function update_attendee( $attendee, $attendee_data ) {
...
+		// Bail early if user is not owner/have permissions.
+		if ( ! $this->user_can_manage_attendees( 0, $event_id ) ) {
+			return;
+		}
+
+		echo sprintf(
+			'<a target="_blank" href="%s" class="export action page-title-action" rel="noopener noreferrer">%s</a>',
+			esc_url( $export_url = $this->get_export_url() ),
+			esc_html__( 'Export', 'event-tickets' ) ) ;
+
```
Comment (paulmskim) [inline src/Tribe/Attendees.php:981]:
```suggestion
		echo sprintf(
			'<a target="_blank" href="%s" class="export action page-title-action" rel="noopener noreferrer">%s</a>',
			esc_url( $export_url = $this->get_export_url() ),
			esc_html__( 'Export', 'event-tickets' )
		) ;
```

File: src/admin-views/attendees.php [hunk 13]
```diff
@@ -39,10 +40,13 @@
 			 * Add an action to render content after text title.
 			 *
 			 * @since 5.1.0
+			 * @since TBD Added the attendees information.
 			 *
 			 * @param int $event_id Post ID.
+			 * @param Tribe__Tickets__Attendees @attendees The attendees object.
```
Comment (juanfra) [inline src/admin-views/attendees.php:46]:
```suggestion
			 * @param Tribe__Tickets__Attendees $attendees The attendees object.
```

**Review by juanfra — CHANGES_REQUESTED** [review]
One last minor thing! We're almost there!

**Review by bordoni — APPROVED** [review]

**Pull request merged**